### PR TITLE
GC: Backup: Exchange: Utilize Internal encode/serialization writer

### PR DESCRIPTION
**Pull request by dadams39**
## Description
To reduce memory usage, use the provided serializer object for serialization steps. 

## Does this PR need a docs update or release note?
- [x] :no_entry: No 

## Type of change

- [x] :sunflower: Feature

## Test Plan

- [x] :zap: Unit test

### CODE REVIEW

File: src/internal/connector/exchange/exchange_data_collection.go [hunk 1]
```diff
@@ -342,8 +341,8 @@ func getModTime(mt modTimer) time.Time {
 // responses into data.Stream items contained within the Collection
 type GraphSerializeFunc func(
 	ctx context.Context,
-	client *msgraphsdk.GraphServiceClient,
-	objectWriter *kioser.JsonSerializationWriter,
+	// TODO: Keepers: Not sure how to have the interface use Serialize without updating all the files.
+	service *graph.Service,
```
Comment (ryanfkeepers) [inline src/internal/connector/exchange/exchange_data_collection.go:345]: This should use the `graph.Servicer` interface, not the `Service` object.  Go idioms: accept interfaces, return structs.  That way you can drop all the type identification.

[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/connector/exchange/exchange_data_collection.go:345]: If you're asking if you should add `Serialize` to the `graph.Servicer` interface: yes, I think that's a reasonable thing to do.

**Comment by ashmrtn** [comment]
+1 to keepers' comments

File: src/internal/connector/exchange/exchange_data_collection.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (dadams39) [reply src/internal/connector/exchange/exchange_data_collection.go]: Resovled within [8a919ea](https://github.com/alcionai/corso/pull/2045/commits/8a919ead958b4613703715bbaacb2718790947f8)

Comment (dadams39) [reply src/internal/connector/exchange/exchange_data_collection.go]: Updated [dd42105](https://github.com/alcionai/corso/pull/2045/commits/dd42105b3e21b897ee2309ae4839869abe1ae4d8)

File: src/internal/connector/sharepoint/data_collections_test.go [hunk 2]
```diff
@@ -41,6 +43,18 @@ type SharePointLibrariesSuite struct {
 	suite.Suite
 }
 
+func (suite *SharePointLibrariesSuite) Client() *msgraphsdk.GraphServiceClient {
```
Comment (ryanfkeepers) [inline src/internal/connector/sharepoint/data_collections_test.go:46]: For separation of concerns, the suite itself shouldn't comply with other interfaces.  In the event that we decide to populate these methods for mocking purposes, having them bound to the suite prohibits per-test control of their behavior.  A `type mockServicer struct {}` would be the preferred owner.

File: src/internal/connector/sharepoint/helper_test.go [hunk 3]
```diff
@@ -12,23 +11,6 @@ import (
 	"github.com/alcionai/corso/src/pkg/account"
 )
 
-// ---------------------------------------------------------------------------
-// SharePoint Test Services
-// ---------------------------------------------------------------------------
-type MockGraphService struct{}
```
[the remark this answers is not in the document]
Comment (dadams39) [reply src/internal/connector/sharepoint/helper_test.go:18]: These were removed in some of the other tests that I inspected at the time. Adding the information to the suite itself allowed the person reading the test to know that these structs had nothing within them.

File: src/cmd/getM365/getItem.go [hunk 4]
```diff
@@ -115,21 +115,22 @@ func runDisplayM365JSON(
...
-	sw := kw.NewJsonSerializationWriter()
-
 	response, err := get(ctx, user, m365ID)
 	if err != nil {
 		return errors.Wrap(err, support.ConnectorStackErrorTrace(err))
 	}
 
 	// First return is the number of bytes that were serialized. Ignored
-	_, err = serializeFunc(ctx, gs.Client(), sw, channel, response, user)
+	service := gs.(graph.Service)
```
[the remark this answers is not in the document]
Comment (dadams39) [reply src/cmd/getM365/getItem.go]: Resolved within [91f2556](https://github.com/alcionai/corso/pull/2045/commits/91f2556e4c538ce02f2d411d2dd403d4f7b5bb55)

File: src/internal/connector/sharepoint/helper_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (dadams39) [reply src/internal/connector/sharepoint/helper_test.go:18]: Resolved within [a1e1ebb](https://github.com/alcionai/corso/pull/2045/commits/a1e1ebb5c32f30458996d1c225ec06fbdab007db)

File: src/internal/connector/sharepoint/data_collections_test.go (hunk 2 above)
Comment (dadams39) [reply src/internal/connector/sharepoint/data_collections_test.go]: Resolved within [a1e1ebb](https://github.com/alcionai/corso/pull/2045/commits/a1e1ebb5c32f30458996d1c225ec06fbdab007db)

File: src/internal/connector/graph/service.go [hunk 5]
```diff
@@ -45,10 +46,15 @@ func (s Service) Client() *msgraphsdk.GraphServiceClient {
...
 }
 
-// Seraialize writes an M365 parsable object into a byte array using the built-in
-// application/json writer within the adapter.
 func (s Service) Serialize(object absser.Parsable) ([]byte, error) {
 	writer, err := s.adapter.GetSerializationWriterFactory().GetSerializationWriter("application/json")
+	defer func() {
+		writerErr := writer.Close()
+		if writerErr != nil {
+			err = support.WrapAndAppend("failure during writer closer", writerErr, err)
```
Comment (ryanfkeepers) [inline src/internal/connector/graph/service.go:54]: This has a gotcha.  even though it updates the `err` value, it won't change the response value.  If you want to return this as an error, you have to name your return values.  Ex: `(bs []byte, err error)`.  Although, imo, that's a bit of an anti-pattern.  I think an info log would be sufficient here.

**Review by ryanfkeepers — APPROVED** [review]
approving to unblock, but the comment I left should get addressed first.

**Comment by dadams39** [comment]
Itemer interface makes much of this code not necessary. Closing accordingly.

**Pull request closed without merging**